### ISSUE TITLE
[FEATURE] Investigate GetXAttr performance

### ISSUE THREAD

**Issue by Kern--** [awslabs/soci-snapshotter#971]
### Description

Looking at CPU profiles for a couple of containers, it seems that `GetXAttr` is consistently showing up as a non-trivial amount of the soci-snapshotter's runtime. Specifically it is `n.IsOpaque()` which checks for the existance of a whiteout in the metadata db:
https://github.com/awslabs/soci-snapshotter/blob/6dcfca7bc0c70b00a5f23813d468379d6fbc8de5/fs/layer/node.go#L288-L293

example profiles:
![cpu_runtime](https://github.com/awslabs/soci-snapshotter/assets/2304727/3800b11a-cd8a-42ba-9053-ea8ac2201ead)
![cpu_runtime](https://github.com/awslabs/soci-snapshotter/assets/2304727/2b6de9a6-3028-4c40-abf6-26afcb9fe757)

It's possible that this is a red herring and what we're actually seeing is just that GetXAttr is a very common, so we should be careful how much effort we put into this before validating that we're seeing benefit.

### Describe the solution you'd like

We should investigate whether there is a way we can reduce the cost of `n.IsOpaque()`. 

We could possibly move the opaqueness to a property of the node itself rather than pushing it down into the metadata db. A good first step might be to try this and compare benchmarks to see if it has an effect.

**Comment by Kern--** [comment]
Copying some context


> Ok, I finally got around to testing how this interacts with OverlayFS. The answer is that OverlayFS handles deleted files by creating "whiteouts" in higher layers. For example, if a lower layer has `dir/file` then a higher layer can delete it by containing a file named `dir/.wh.file`. This works with the change.
> 
> For directories, overlayfs has an "opaque" xattr (`trusted.overlay.opaque=y`) which treats all files as deleted rather than having to explicitly create a whiteout file for each. In containers, this is encoded as a special `.wh..wh..opq` file in a higher layer (.e.g. to delete all files in `/bin` a higher layer would contain `/bin/.wh..wh..opq`). So it's possible for the container image to contain an opaque directory with no xattrs. This change will disable xattrs and the directory will be populated at runtime.
> 
> An example:
> 
> ```
> FROM alpine:latest
> 
> RUN mkdir /dir && touch /dir/file1 /dir/file2 /dir/file3
> RUN touch /dir/.wh..wh..opq
> ```
> 
> Without Soci:
> 
> ```
> # ls /dir
> #
> ```
> 
> With SOCI:
> 
> ```
> # ls /dir
> file1 file2 file3
> #
> ```
> 
> The [image spec](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc5/layer.md#whiteouts) says that whiteouts should be explicit, but to be spec compliant we must accept opaque directories. At a minimum, we should look for `.wh..wh..opq` files when determining if a layer has xattrs.

https://github.com/awslabs/soci-snapshotter/pull/718#issuecomment-1875759931

> Taking a step back and considering #971, I think the real issue we have is that checking opaqueness is expensive. Each check goes through bbolt to see if a file exists. Since the lower layers are immutable, I think we could probably just push opaqueness up into the in-memory nodes.
> 
> I'm going to run some quick hard-coded benchmarks (normal, disableXattrs hard coded to true, isOpaque hard coded to false) to see how that compares.


https://github.com/awslabs/soci-snapshotter/pull/718#issuecomment-1875766120

**Comment by sondavidb** [comment]
TL;DR GetXAttrs is an expensive operation and should be avoided unless necessary. Full testing and results on changes can be found here: https://github.com/awslabs/soci-snapshotter/pull/1021#issuecomment-2010966698

Operations that normally made significant calls to GetXAttrs would take up a nontrivial amount of CPU seconds due to checking if `isOpaque` is true. #1021 fixed this by not making these calls on a layer-level unless the directory had opaque directories.

While we may want further iteration of this (i.e. see if we can make it more granular to avoid the call more often), the question and immediate problem at hand have been resolved, so this can be closed.
